packages/core/contracts/optimistic-assertor/implementation/examples/DataAsserter.sol

```diff
@@ -0,0 +1,104 @@
+        if (!assertionsData[dataAssertionId].resolved) return (false, 0);
+        return (true, assertionsData[dataAssertionId].data);
+    }
+
+    // Asserts data for a specific dataId on behalf of an asserter address.
+    function assertDataFor(
+        bytes32 dataId,
+        bytes32 data,
+        address asserter
+    ) public returns (bytes32 oaAssertionId) {
+        bytes32 dataAssertionId = getAssertionId(dataId, asserter);
+        asserter = asserter == address(0) ? msg.sender : asserter;
+        require(assertionsData[dataAssertionId].asserter == address(0), "Data already asserted");
+        uint256 bond = oa.getMinimumBond(address(defaultCurrency));
+        defaultCurrency.safeTransferFrom(msg.sender, address(this), bond);
+        defaultCurrency.safeApprove(address(oa), bond);
+        oaAssertionId = oa.assertTruthFor(
+            abi.encodePacked(
+                "Data asserted for dataId: 0x",
+                AncillaryData.toUtf8Bytes(dataId),
+                " and asserter: 0x",
+                AncillaryData.toUtf8BytesAddress(asserter),
+                " at timestamp: ",
+                AncillaryData.toUtf8BytesUint(block.timestamp),
+                " in the DataAsserter contract at 0x",
+                AncillaryData.toUtf8BytesAddress(address(this)),
+                " is valid."
+            ),
+            asserter,
+            address(this),
+            address(0), // No sovereign security manager.
+            defaultCurrency,
+            bond,
+            assertionLiveness,
+            defaultIdentifier
+        );
+        assertionsData[dataAssertionId] = DataAssertion(dataId, data, asserter, false);
+        oaIdsToInternalIds[oaAssertionId] = dataAssertionId;
+        emit DataAsserted(dataId, data, asserter, oaAssertionId);
+    }
+
+    // OptimisticAssertor callback.
+    function assertionResolved(bytes32 assertionId, bool assertedTruthfully) public {
+        require(msg.sender == address(oa));
+        // If the assertion was true, then the data assertion is resolved.
+        if (assertedTruthfully) {
+            assertionsData[oaIdsToInternalIds[assertionId]].resolved = true;
+            DataAssertion memory dataAssertion = assertionsData[oaIdsToInternalIds[assertionId]];
+            emit DataAssertionResolved(dataAssertion.dataId, dataAssertion.data, dataAssertion.asserter, assertionId);
+            // Else delete the data assertion if it was false so the same asserter can assert it again.
+        } else delete assertionsData[oaIdsToInternalIds[assertionId]];
+    }
+
+    // If assertion is disputed, do nothing and wait for resolution.
+    // This OptimisticAssertor callback function needs to be defined so the OA doesn't revert when it tries to call it.
+    function assertionDisputed(bytes32 assertionId) public {}
+
+    // Returns the internal ID for a data assertion.
+    function getAssertionId(bytes32 dataId, address asserter) public pure returns (bytes32) {
```

Comment:
one question I have is _all_ assertions in this contract make a call to the OA on `assertTruthFor`. for this reason, why do you think we need both an internal notion of a _data_ assertion and a OA assertion? would it not be posible to do everything via the OA assertion identifier?

Reply:
This is an excellent observation. I did it this way so that the consumer contract could read from the Data Asserter using its own ids without storing the oaAssertionId. However, I believe it is much more logical to simplify it as you advise and remove the unnecessary id I made

Fixed [here](https://github.com/UMAprotocol/protocol/pull/4261/commits/971e1c21c58c74383ddef17425c8ac36b731e90e)

Reply:
note that this changes contract behavior as the same dataId/asserter combination is no more blocking new assertions due to having different claims (timestamp is added)